File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/queen-attack/queen-attack.js
//
// This is only a SKELETON file for the 'Queen Attack' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class QueenAttack {
  constructor({
    black: [blackRow, blackColumn] = [7, 3],
    white: [whiteRow, whiteColumn] = [0, 3],
  } = {}) {
    if (
      blackRow < 0 || blackRow > 7 ||
      blackColumn < 0 || blackColumn > 7 ||
      whiteRow < 0 || whiteRow > 7 ||
      whiteColumn < 0 || whiteColumn > 7
    ) {
      throw new Error('Queen must be placed on the board');
    }

    if (blackRow === whiteRow && blackColumn === whiteColumn) {
      throw new Error('Queens cannot share the same space');
    }

    this.black = [blackRow, blackColumn];
    this.white = [whiteRow, whiteColumn];
  }

  toString() {
    const board = Array.from({length: 8}, () => Array(8).fill('_'));
    board[this.black[0]][this.black[1]] = 'B';
    board[this.white[0]][this.white[1]] = 'W';
    return board.reverse()
      .map(row => row.join(' '))
      .join('\n') + '\n';
  }

  get canAttack() {
    const [br, bc] = this.black;
    const [wr, wc] = this.white;
    return br === wr || bc === wc || Math.abs(br - wr) === Math.abs(bc - wc);
  }
}

File: src/2025-02-24-deepclaude-rust/cpp/exercises/practice/queen-attack/queen_attack.cpp
#include "queen_attack.h"
#include <cmath>
#include <stdexcept>

namespace queen_attack {

chess_board::chess_board(std::pair<int, int> white, std::pair<int, int> black) 
    : white_(white), black_(black) {
    if (white == black) {
        throw std::domain_error("Queens cannot occupy the same position.");
    }
    if (white.first < 0 || white.first >= 8 || white.second < 0 || white.second >= 8 ||
        black.first < 0 || black.first >= 8 || black.second < 0 || black.second >= 8) {
        throw std::out_of_range("Queen position out of bounds.");
    }
}

bool chess_board::can_attack() const {
    return white_.first == black_.first || 
           white_.second == black_.second ||
           std::abs(white_.first - black_.first) == std::abs(white_.second - black_.second);
}

}  // namespace queen_attack
